fxa-client/sdks/android/library/src/main/java/org/mozilla/fxa_client/internal/RustObject.kt

```diff
@@ -0,0 +1,87 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fxa_client.internal
+
+import com.sun.jna.Pointer
+import com.sun.jna.PointerType
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Base class that wraps an non-optional [Pointer] representing a pointer to a Rust object.
+ * This class implements [AutoCloseable] but does not provide an implementation, forcing all
+ * subclasses to implement it.
+ */
+abstract class RustObject<T: PointerType> internal constructor(pointer: T? = null) : AutoCloseable {
+    private var rawPointer: AtomicReference<T?> = AtomicReference(pointer)
+
+    val isConsumed: Boolean
+        get() = this.rawPointer.get() == null
+
+    fun validPointer(): T {
+        // TODO: It feels like this should throw something more specific than NullPointerException...
+        return this.rawPointer.get()!!
+    }
+
+    fun consumePointer(): T {
+        return this.rawPointer.getAndSet(null)!!
+    }
+
+    protected abstract fun destroy(p: T)
+
+    override fun close() {
+        destroy(consumePointer())
+    }
+
+    internal inline fun <U> rustCall(callback: (Error.ByReference) -> U?): U {
+        // Not sure if I can use @Synchronized in an `inline fun`
+        return synchronized(this) {
+            unlockedRustCall(callback)
+        }
+    }
+
+    internal inline fun <U> nullableRustCall(callback: (Error.ByReference) -> U?): U? {
+        return synchronized(this) {
+            unlockedNullableRustCall(callback)
+        }
+    }
+}
+
+internal inline fun <U> unlockedNullableRustCall(callback: (Error.ByReference) -> U?): U? {
+    val e = Error.ByReference()
+    try {
+        val ret = callback(e)
+        if (e.isFailure()) {
+            throw FxaException.fromConsuming(e)
+        }
+        return ret
+    } finally {
+        // This only matters if `callback` throws or does a non-local return, which
+        // we currently don't do.
+        e.ensureConsumed()
+    }
+}
+
+// The vast majority of our functions return null only on errors, while writing
+// information about the error to the error output. Rather than have each caller do
+// the `!!` that indicates no error occurred, we just do it here.
+internal inline fun <U> unlockedRustCall(callback: (Error.ByReference) -> U?): U {
```

Comment:
Could we add a little bit of doc saying explaining the different between `unlockedRustCall` and `rustCall`?
Thinking about it we should probably add documentation for all the methods in `RustObject` (we can do it later), because it might no be obvious in a year or two down the line what this class does.